### PR TITLE
Expose a torch.result_type and simplify tensor iterator

### PR DESCRIPTION
**Pull request by nairbv**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26655 Add torch.promote_types function
* #26051 Add type promotion support in sparse tensors
* **#26012 Expose a torch.result_type and simplify tensor iterator**

Differential Revision: [D17556197](https://our.internmc.facebook.com/intern/diff/D17556197)

### CODE REVIEW

File: aten/src/ATen/core/OpsAlreadyMovedToC10.cpp [hunk 1]
```diff
@@ -1098,6 +1098,9 @@ bool aten_op_is_not_moved_to_c10_yet(const c10::OperatorName& opName) {
         {"aten::to", "dtype_layout"},
         {"aten::to", "device"},
         {"aten::to", "dtype"},
+        {"aten::result_type", ""},
```
Comment (gchanan) [inline aten/src/ATen/core/OpsAlreadyMovedToC10.cpp]: why is this here 3 times?

File: torch/csrc/autograd/utils/wrap_outputs.h [hunk 2]
```diff
@@ -47,6 +48,10 @@ inline PyObject* wrap(THPDtype *dtype) {
   return (PyObject*)dtype;
 }
 
+inline PyObject* wrap(at::ScalarType scalarType) {
```
Comment (gchanan) [inline torch/csrc/autograd/utils/wrap_outputs.h]: seems like if we have all this we can probably port promoteTypes to native_functions and this would just work.  Are you planning to do that?

File: torch/_torch_docs.py [hunk 3]
```diff
@@ -4431,6 +4431,26 @@ def merge_dicts(*dicts):
 """)
 
 
+add_docstr(torch.result_type,
+           r"""
+result_type(tensor, other) -> ScalarType
```
Comment (gchanan) [inline torch/_torch_docs.py]: this should be python first -- ScalarType is a C++ concept.

File: torch/_torch_docs.py [hunk 4]
```diff
@@ -4431,6 +4431,26 @@ def merge_dicts(*dicts):
...
+
+Returns the :class:`torch.dtype` that would result from performing an arithmetic
+operation on the provided input tensors.
+
+Args:
+    tensor (Tensor): an input tensor
```
Comment (gchanan) [inline torch/_torch_docs.py]: can both arguments not be Tensors or Number (or scalar, I don't know how we do this consistently in documentation).

Also, can you link this to the explanation of type promotion we have elsewhere in the docs?

File: aten/src/ATen/native/native_functions.yaml [hunk 5]
```diff
@@ -3518,6 +3518,15 @@
   variants: method
   named_guard: False
 
+- func: result_type(Tensor tensor, Tensor other) -> ScalarType
```
Comment (gchanan) [inline aten/src/ATen/native/native_functions.yaml]: how did you decide on these overloads?  For example, I'm a bit surprised this doesn't follow the `pow` pattern.

**Review by gchanan — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/native_functions.yaml [hunk 6]
```diff
@@ -3518,6 +3518,15 @@
...
+  variants: function
+
+- func: result_type(Tensor tensor, Scalar other) -> ScalarType
```
Comment (gchanan) [inline aten/src/ATen/native/native_functions.yaml]: so...it's not actually enforced that we have overload names?  @dzhulgakov / @smessmer is your current story that we just don't need to worry about it?

File: aten/src/ATen/core/ivalue.h [hunk 7]
```diff
@@ -186,6 +187,10 @@ struct CAFFE2_API IValue final {
     return payload.as_double;
   }
 
+  IValue(ScalarType t) : tag(Tag::ScalarType) {
```
Comment (gchanan) [inline aten/src/ATen/core/ivalue.h]: nothing else in this file is constructed in this way -- they all set is_intrusive_ptr(false) as well.  Why the difference here?

File: aten/src/ATen/core/ivalue.h [hunk 8]
```diff
@@ -186,6 +187,10 @@ struct CAFFE2_API IValue final {
...
+    payload.as_int = static_cast<std::underlying_type<ScalarType>::type>(t);
+  }
+
```
Comment (gchanan) [inline aten/src/ATen/core/ivalue.h]: naively, I would also expect an isScalarType and that toScalarType actually checks the tag is a scalar type.

I'm admittedly not sure if we need to deal with BC constraints here, though.  @suo do you know?

**Review by gchanan — CHANGES_REQUESTED** [review]

File: aten/src/ATen/core/OpsAlreadyMovedToC10.cpp (hunk 1 above)
Comment (gchanan) [reply aten/src/ATen/core/OpsAlreadyMovedToC10.cpp]: rather, 2 times.  oh, I guess this is related to the name overloads, which is discussed below.

File: aten/src/ATen/native/TypeProperties.cpp [hunk 9]
```diff
@@ -44,4 +44,63 @@ Tensor type_as(const Tensor& self, const Tensor& other) {
...
+  if (higher == ScalarType::Bool || isFloatingType(lower)) {
+    return promote_skip_undefined(higher, lower);
+  }
+  if (higher != ScalarType::Undefined) {
+      return higher;
+  }
+  return lower;
+}
+
+ScalarType _result_type(TensorList tensors) {
```
Comment (gchanan) [inline aten/src/ATen/native/TypeProperties.cpp]: why is this declared as a native_function if it's just a helper?

File: torch/csrc/autograd/utils/wrap_outputs.h (hunk 2 above)
Comment (nairbv) [reply torch/csrc/autograd/utils/wrap_outputs.h]: yes, but in a separate PR

File: aten/src/ATen/core/OpsAlreadyMovedToC10.cpp (hunk 1 above)
Comment (nairbv) [reply aten/src/ATen/core/OpsAlreadyMovedToC10.cpp]: this is a generated file so I'm not entirely sure, but yeah seems to be due to overloads

File: aten/src/ATen/native/TypeProperties.cpp (hunk 9 above)
Comment (nairbv) [reply aten/src/ATen/native/TypeProperties.cpp]: I want to call this from TensorIterator. If it's only in the .cpp file it isn't accessible except within this file. I believe that's because native_functions causes .h files to be generated.

File: aten/src/ATen/core/ivalue.h (hunk 7 above)
Comment (nairbv) [reply aten/src/ATen/core/ivalue.h]: yup this was a mistake

File: aten/src/ATen/core/ivalue.h (hunk 8 above)
Comment (nairbv) [reply aten/src/ATen/core/ivalue.h]: this was wrong, will fix

File: aten/src/ATen/native/native_functions.yaml (hunk 5 above)
Comment (nairbv) [reply aten/src/ATen/native/native_functions.yaml]: My justification for this was based on `eq`, which supports `torch.eq(tensor, tensor)` or `torch.eq(tensor, scalar)` but not `torch.eq(scalar, tensor)`.

It's not ideal. I could add all possible combinations of overloads, but that seems messy. I assume there's a better way of doing this since it works both ways for `add` but I haven't investigated how that works. Would you be OK with extending this to support `(Scalar, Tensor)` and `(Scalar, Scalar)` in a separate PR?

File: torch/_torch_docs.py (hunk 4 above)
Comment (nairbv) [reply torch/_torch_docs.py]: it's in the linked torch.dtype doc but added an explicit reference.

File: aten/src/ATen/core/ivalue.h (hunk 8 above)
Comment (nairbv) [reply aten/src/ATen/core/ivalue.h]: added the is_intrusive_ptr, isScalarType, and added the check in toScalarType

File: aten/src/ATen/core/OpsAlreadyMovedToC10.cpp (hunk 1 above)
Comment (gchanan) [reply aten/src/ATen/core/OpsAlreadyMovedToC10.cpp]: for now I'd just add name overloads using whatever pattern you find in the file (maybe use pow as an example?)

File: aten/src/ATen/native/native_functions.yaml (hunk 5 above)
Comment (gchanan) [reply aten/src/ATen/native/native_functions.yaml]: ya, separate PR sounds fine.  It also seems fine to not add this, assuming pow doesn't do type promotion (otherwise, you couldn't use this function to test pow).

File: aten/src/ATen/native/TypeProperties.cpp (hunk 9 above)
Comment (gchanan) [reply aten/src/ATen/native/TypeProperties.cpp]: just declare it in a .h file and include that .h file from TensorIterator.

File: torch/_torch_docs.py [hunk 10]
```diff
@@ -4431,6 +4431,27 @@ def merge_dicts(*dicts):
 """)
 
 
+add_docstr(torch.result_type,
+           r"""
+result_type(tensor, other) -> dtype
+
+Returns the :class:`torch.dtype` that would result from performing an arithmetic
+operation on the provided input tensors. See :class:`torch.dtype` documentation
```
Comment (gchanan) [inline torch/_torch_docs.py]: can we link directly to that specific section?

File: aten/src/ATen/core/ivalue.h (hunk 8 above)
Comment (gchanan) [reply aten/src/ATen/core/ivalue.h]: ok, this looks correct now, but I don't know about the BC concerns.

Comment (nairbv) [reply aten/src/ATen/core/ivalue.h]: what is it that would potentially backwards incompatible?

Comment (gchanan) [reply aten/src/ATen/core/ivalue.h]: my concern was you serialized something as a tagged-int previously and when you load it in the new version it errors out.

File: aten/src/ATen/native/native_functions.yaml (hunk 5 above)
Comment (nairbv) [reply aten/src/ATen/native/native_functions.yaml]: added all combinations (including scalar scalar which pow doesn't have because add does).

File: torch/_torch_docs.py (hunk 10 above)
Comment (nairbv) [reply torch/_torch_docs.py]: added a reference in the doc and a direct link to it

**Review by gchanan — APPROVED** [review]
non-JIT parts look good -- I'd defer to @suo or others on the JIT parts.

File: aten/src/ATen/core/ivalue.h (hunk 8 above)
Comment (nairbv) [reply aten/src/ATen/core/ivalue.h]: hmm I do have a test failing here now.
`python test_cpp_api_parity.py TestCppApiParity.test_torch_nn_SampleModule_has_parity` gives "ivalue not a ScalarType" and I assume that's because it's trying to deserialize an int as a scalar type.

Comment (nairbv) [reply aten/src/ATen/core/ivalue.h]: removed the isScalarType check. toLayout, toMemoryFormat, etc don't have it and it was causing test failures.

Comment (zdevito) [reply aten/src/ATen/core/ivalue.h]: For correctness, adding a new tag requires adding a corresponding type in jit_type.h, and adjusting the schema/type parsers to understand that 'ScalarType' maps to this type. This PR seems to just add the tag and not add a type. If we need to keep the tag around, then we should also add the corresponding type and fix the type parser.

Comment (nairbv) [reply aten/src/ATen/core/ivalue.h]: @zdevito this PR does add a new type in jit_type.h, is it incorrect?

Comment (nairbv) [reply aten/src/ATen/core/ivalue.h]: https://github.com/pytorch/pytorch/pull/26012/files#diff-7e8e36dc4eff19e3d5d1daa977da20eeR1142

Comment (nairbv) [reply aten/src/ATen/core/ivalue.h]: removed almost all jit changes

**Comment by nairbv** [comment]
related issue: https://github.com/pytorch/pytorch/issues/25472
